ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,175 @@
+                                                    -> {fldname; stat={created;modified;deleted}; value=value}::acc) obj [] in
+  let handle_table table = Table.fold_over_recent token (fun objref {created; modified; deleted} value acc
+                                                          -> {objref; stat={created;modified;deleted}; value=(handle_obj value)}::acc) table [] in
+  TableSet.fold_over_recent token (fun tblname {created; modified; deleted} value acc
+                                    -> {tblname; stat={created;modified;deleted}; value=(handle_table value)}::acc) ts []
+
+let get_delta ~__context ~token =
+  let db = Db_ref.get_database (Context.database_of __context) in
+  let db_access =
+    fun () -> Mutex.execute Db_lock.dbcache_mutex (fun () ->
+        {fresh_token=Manifest.generation (Database.manifest db);
+         tables=(check_for_updates token db);
+         deletes=(get_deleted __context token db)})
+  in
+  let _dbreply = db_access () in
+  _dbreply
+
+let handler (req: Request.t) s _ =
+  req.Http.Request.close <- true;
+  let token = Generation.of_string (snd (List.find (fun (x,y) -> x = "token") req.query)) in
+  Xapi_http.with_context "Database backup" req s (fun __context ->
+      let d = get_delta __context token in
+      let str_update = (Jsonrpc.to_string (rpc_of_delta d)) in
+      Http_svr.response_str req s str_update
+    )
+
+(* Slave Functions *)
+
+let send_notification (c:Context.t) reason p =
+  let record = (Eventgen.find_get_record p.tbl ~__context:c ~self:p.row) () in
+  begin match record with
+    | None -> debug "Could not send %s event for %s in %s table" reason p.row p.tbl;
+    | Some record -> debug "Sending %s event for %s object in %s table" reason p.row p.tbl;
+      Db_action_helper.events_notify ~snapshot:record p.tbl reason p.row;
+  end
+
+let apply_deletes __context (deletes:del_tables list) =
+  List.iter
+    (fun dtbl ->
+       (fun tblname objref ->
+          if Db_cache_impl.is_valid_ref (Context.database_of __context) objref then
+            Xapi_slave_db.slave_db := Db_cache_types.remove_row tblname objref !Xapi_slave_db.slave_db
+          else
+            Printf.printf "Not valid ref"
+       ) dtbl.table dtbl.key
+    ) deletes
+
+let make_change __context mtime tblname objref fldname newval =
+  Xapi_slave_db.slave_db :=
+    Database.update (
+      (fun _ -> newval)
+      |> Row.update mtime fldname newval
+      |> Table.update mtime objref Row.empty
+      |> TableSet.update mtime tblname Table.empty) !Xapi_slave_db.slave_db
+
+let unpack_row __context (table:table) tblname row =
+  List.iter (fun (f:field)->
+      Mutex.execute Xapi_slave_db.slave_db_mutex
+        (fun () -> make_change __context f.stat.modified tblname row.objref f.fldname f.value)
+    ) row.value
+
+let unpack_table __context (t:table) =
+  List.iter (unpack_row __context t t.tblname) (if t.value = [] then [empty_row] else t.value)
```

Comment:
I still don't really follow the `empty_row` thing here. Can't we just have a table without rows?